pmd-core/src/main/java/net/sourceforge/pmd/processor/AbstractPMDProcessor.java

```diff
@@ -108,25 +115,44 @@ protected RuleSets createRuleSets(RuleSetFactory factory, Report report) {
     @SuppressWarnings("PMD.CloseResource")
     // the data sources must only be closed after the threads are finished
     // this is done manually without a try-with-resources
+    @Deprecated
     public void processFiles(RuleSetFactory ruleSetFactory, List<DataSource> files, RuleContext ctx,
-            List<Renderer> renderers) {
+                             List<Renderer> renderers) {
         try {
             final RuleSets rs = createRuleSets(ruleSetFactory, ctx.getReport());
-            configuration.getAnalysisCache().checkValidity(rs, configuration.getClassLoader());
+            processFiles(rs, files, ctx, renderers);
+        } finally {
+            // in case we analyzed files within Zip Files/Jars, we need to close them after
+            // the analysis is finished
+            for (DataSource dataSource : files) {
+                IOUtils.closeQuietly(dataSource);
+            }
+        }
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    // the data sources must only be closed after the threads are finished
+    // this is done manually without a try-with-resources
+    public void processFiles(RuleSets rulesets, List<DataSource> files, RuleContext ctx, List<Renderer> renderers) {
+        try {
+            reportBrokenRules(ctx.getReport(), rulesets);
+            configuration.getAnalysisCache().checkValidity(rulesets, configuration.getClassLoader());
             final SourceCodeProcessor processor = new SourceCodeProcessor(configuration);
 
             for (final DataSource dataSource : files) {
                 // this is the real, canonical and absolute filename (not shortened)
                 String realFileName = dataSource.getNiceFileName(false, null);
 
-                runAnalysis(new PmdRunnable(dataSource, realFileName, renderers, ctx, rs, processor));
+                runAnalysis(new PmdRunnable(dataSource, realFileName, renderers, ctx, rulesets, processor));
             }
 
             // render base report first - general errors
             renderReports(renderers, ctx.getReport());
 
             // then add analysis results per file
             collectReports(renderers);
+        } catch (RuntimeException e) {
+            throw new ContextedRuntimeException(e).addContextValue("filename", String.valueOf(ctx.getSourceCodeFile()));
```

Comment:
That's an interesting information. At some point, we should define that with a specific exception.

But: do we at this position actually know, what file we were analyzing? This is the outer catch, so ctx is maybe the last one. And in case of multi threading, the ctx here is not used at all.... or do I miss something here?

Runtime Exception are already handled in PmdRunnable and added there as errors to the report... So, I'm not sure, which RuntimeExceptions we would get here...

Reply:
I introduced that as a workaround for PMDTask, because the new entry point hides its RuleContext. Before, PMDTask was querying the RuleContext when there was an exception. So, I don't know if this makes sense, but it should match the current behavior. We can do better in PMD7. I'd like to introduce FileAnalysisException, which is a supertype for eg TokenMgrError and ParseException, and has a file name.

Reply:
Hm... Maybe PMDTask was/is already wrong?
FileAnalysisException sounds good.